rules.d/50-udev-default.rules.in

```diff
@@ -54,6 +54,24 @@ SUBSYSTEM=="firewire", ATTR{units}=="*0x00b09d:0x00010*", GROUP="video"
 SUBSYSTEM=="firewire", ATTR{units}=="*0x00a02d:0x010001*", GROUP="video"
 SUBSYSTEM=="firewire", ATTR{units}=="*0x00a02d:0x014001*", GROUP="video"
 
+# Drivers in ALSA firewire stack expects userspace application to control target device.
+SUBSYSTEM=="firewire", ATTR{vendor}=="0x001486", ATTR{units}=="0x00a02d:0x010000", GROUP="audio", ENV{FW_AUDIO_AND_MUSIC}="1"
+SUBSYSTEM=="firewire", ATTR{vendor}=="0x000ff2", ATTR{units}=="0x00a02d:0x010000", GROUP="audio", ENV{FW_AUDIO_AND_MUSIC}="1"
+SUBSYSTEM=="firewire", ATTR{vendor}=="0x00075b", ATTR{units}=="0x00a02d:0x010000", GROUP="audio", ENV{FW_AUDIO_AND_MUSIC}="1"
+SUBSYSTEM=="firewire", ATTR{guid}=="0x??????04????????", ATTR{units}=="0x??????:0x000001", GROUP="audio", ENV{FW_AUDIO_AND_MUSIC}="1"
+SUBSYSTEM=="firewire", ATTR{guid}=="0x??????00????????", ATTR{units}=="0x??????:0x000001", GROUP="audio", ENV{FW_AUDIO_AND_MUSIC}="1"
+SUBSYSTEM=="firewire", ATTR{guid}=="0x??????10????????", ATTR{units}=="0x??????:0x000001", GROUP="audio", ENV{FW_AUDIO_AND_MUSIC}="1"
+SUBSYSTEM=="firewire", ATTR{guid}=="0x??????20????????", ATTR{units}=="0x??????:0x000001", GROUP="audio", ENV{FW_AUDIO_AND_MUSIC}="1"
+SUBSYSTEM=="firewire", ATTR{guid}=="0x??????5[12]????????", ATTR{units}=="0x??????:0x000001", GROUP="audio", ENV{FW_AUDIO_AND_MUSIC}="1"
+SUBSYSTEM=="firewire", ATTR{vendor}=="0x00a07e", ATTR{units}=="0x0000a?:0x000001", GROUP="audio", ENV{FW_AUDIO_AND_MUSIC}="1"
+SUBSYSTEM=="firewire", ATTR{vendor}=="0x000022", ATTR{units}=="0x000022:0x80000?", GROUP="audio", ENV{FW_AUDIO_AND_MUSIC}="1"
+SUBSYSTEM=="firewire", ATTR{vendor}=="0x0001f2", ATTR{units}=="0x0001f2:0x0000??", GROUP="audio", ENV{FW_AUDIO_AND_MUSIC}="1"
+SUBSYSTEM=="firewire", ATTR{vendor}=="0x000a35", ATTR{units}=="0x000a35:0x00000[1-5]", GROUP="audio", ENV{FW_AUDIO_AND_MUSIC}="1"
+SUBSYSTEM=="firewire", ATTR{vendor}=="0x00a0de", ATTR{units}=="0x00a0de:0xffffff", GROUP="audio", ENV{FW_AUDIO_AND_MUSIC}="1"
+SUBSYSTEM=="firewire", ATTR{vendor}=="0x000a92", ATTR{units}=="0x00a0de:0xffffff", GROUP="audio", ENV{FW_AUDIO_AND_MUSIC}="1"
+SUBSYSTEM=="firewire", ATTR{vendor}=="0x00a0de", ATTR{units}=="0x00a02d:0x000300", GROUP="audio", ENV{FW_AUDIO_AND_MUSIC}="1"
+SUBSYSTEM=="firewire", ATTR{vendor}=="0x00130e", ATTR{units}=="0x00130e:0x000[27]00", GROUP="audio", ENV{FW_AUDIO_AND_MUSIC}="1"
```

Comment:
please move these into hwdb too!

anything that matches vid/pid should be in hwdb

Reply:
Hi Lennert,

> please move these into hwdb too!
> anything that matches vid/pid should be in hwdb

Yep. It's cool idea and I've investigated it. However, I have a reason specific to IEEE 1394 bus and Linux FireWire subsystem.

Nodes in IEEE 1394 bus includes several units:

```
 * node A
   * unit A-1
   * unit A-2
 * node B
   * unit B-1
```

Linux FireWire subsystem adds devices for both of node and units, then expose cdev just for node:

```
 * node A - (/dev/fw0)
   * unit A-1
   * unit A-2
 * node B (/dev/fw1)
   * unit B-1
```

The device for unit is just exposed to kernel modules bound according to modalias:

```
 * node A - (/dev/fw0)
   * unit A-1 (modalias)
   * unit A-2 (modalias)
 * node B (/dev/fw1)
   * unit B-1 (modalias)
```

As a consequence, the device for node has no modalias:

```
$ udevadm info -a
...
P: /devices/pci0000:00/0000:00:01.2/0000:03:00.2/0000:04:07.0/0000:0a:00.0/0000:0b:00.0/fw1
N: fw1
L: 0
E: DEVPATH=/devices/pci0000:00/0000:00:01.2/0000:03:00.2/0000:04:07.0/0000:0a:00.0/0000:0b:00.0/fw1
E: SUBSYSTEM=firewire
E: DEVNAME=/dev/fw1
E: MAJOR=510
E: MINOR=1
E: USEC_INITIALIZED=435654494
E: ID_PATH=pci-0000:0b:00.0
E: ID_PATH_TAG=pci-0000_0b_00_0
E: ID_FOR_SEAT=firewire-pci-0000_0b_00_0
E: TAGS=:uaccess:seat:

P: /devices/pci0000:00/0000:00:01.2/0000:03:00.2/0000:04:07.0/0000:0a:00.0/0000:0b:00.0/fw1/fw1.0
L: 0
E: DEVPATH=/devices/pci0000:00/0000:00:01.2/0000:03:00.2/0000:04:07.0/0000:0a:00.0/0000:0b:00.0/fw1/fw1.0
E: SUBSYSTEM=firewire
E: MODALIAS=ieee1394:ven0000022Emo00800007sp0000A02Dver00010001
```

```
$ udevadm info -a /sys/devices/pci0000:00/0000:00:01.2/0000:03:00.2/0000:04:07.0/0000:0a:00.0/0000:0b:00.0/fw1/fw1.0
  ...
  looking at device '/devices/pci0000:00/0000:00:01.2/0000:03:00.2/0000:04:07.0/0000:0a:00.0/0000:0b:00.0/fw1/fw1.0':
    KERNEL=="fw1.0"
    SUBSYSTEM=="firewire"
    DRIVER==""
    ATTR{model}=="0x800007"
    ATTR{model_name}=="FireOne"
    ATTR{power/async}=="disabled"
    ATTR{power/control}=="auto"
    ATTR{power/runtime_active_kids}=="0"
    ATTR{power/runtime_active_time}=="0"
    ATTR{power/runtime_enabled}=="disabled"
    ATTR{power/runtime_status}=="unsupported"
    ATTR{power/runtime_suspended_time}=="0"
    ATTR{power/runtime_usage}=="0"
    ATTR{rom_index}=="12"
    ATTR{specifier_id}=="0x00a02d"
    ATTR{version}=="0x010001"

  looking at parent device '/devices/pci0000:00/0000:00:01.2/0000:03:00.2/0000:04:07.0/0000:0a:00.0/0000:0b:00.0/fw1':
    KERNELS=="fw1"
    SUBSYSTEMS=="firewire"
    DRIVERS==""
    ATTRS{guid}=="0x00022e0200000007"
    ATTRS{is_local}=="0"
    ATTRS{model}=="0x800007"
    ATTRS{model_name}=="FireOne"
    ATTRS{power/async}=="disabled"
    ATTRS{power/control}=="auto"
    ATTRS{power/runtime_active_kids}=="0"
    ATTRS{power/runtime_active_time}=="0"
    ATTRS{power/runtime_enabled}=="disabled"
    ATTRS{power/runtime_status}=="unsupported"
    ATTRS{power/runtime_suspended_time}=="0"
    ATTRS{power/runtime_usage}=="0"
    ATTRS{units}=="0x00a02d:0x010001"
    ATTRS{vendor}=="0x00022e"
    ATTRS{vendor_name}=="TASCAM"

  looking at parent device '/devices/pci0000:00/0000:00:01.2/0000:03:00.2/0000:04:07.0/0000:0a:00.0/0000:0b:00.0':
    KERNELS=="0000:0b:00.0"
    SUBSYSTEMS=="pci"
    DRIVERS=="firewire_ohci"
    ATTRS{ari_enabled}=="0"
    ATTRS{broken_parity_status}=="0"
    ATTRS{class}=="0x0c0010"
    ATTRS{consistent_dma_mask_bits}=="32"
    ATTRS{d3cold_allowed}=="1"
    ATTRS{device}=="0x823f"
    ATTRS{dma_mask_bits}=="32"
    ATTRS{driver_override}=="(null)"
    ATTRS{enable}=="1"
    ATTRS{irq}=="36"
    ATTRS{local_cpulist}=="0-7"
    ATTRS{local_cpus}=="000000ff"
    ATTRS{msi_bus}=="1"
    ATTRS{numa_node}=="-1"
    ATTRS{power/async}=="enabled"
    ATTRS{power/control}=="on"
    ATTRS{power/runtime_active_kids}=="0"
    ATTRS{power/runtime_active_time}=="538146"
    ATTRS{power/runtime_enabled}=="forbidden"
    ATTRS{power/runtime_status}=="active"
    ATTRS{power/runtime_suspended_time}=="0"
    ATTRS{power/runtime_usage}=="2"
    ATTRS{power/wakeup}=="disabled"
    ATTRS{power/wakeup_abort_count}==""
    ATTRS{power/wakeup_active}==""
    ATTRS{power/wakeup_active_count}==""
    ATTRS{power/wakeup_count}==""
    ATTRS{power/wakeup_expire_count}==""
    ATTRS{power/wakeup_last_time_ms}==""
    ATTRS{power/wakeup_max_time_ms}==""
    ATTRS{power/wakeup_total_time_ms}==""
    ATTRS{revision}=="0x01"
    ATTRS{subsystem_device}=="0x823f"
    ATTRS{subsystem_vendor}=="0x104c"
    ATTRS{vendor}=="0x104c"
```

As long as I understand, modalias is a key to match in hwdb database. For the way, we just match the device corresponding to the unit. However, permission arrangement is required to the cdev corresponding to the node.

Reply:
modalias are typically used as keys for hwdb, but they don#t have to be. You may build your own keys. There are plenty of examples for that in rules.d/, for example `60-evdev.rules` and `60-sensor.rules`

Reply:
> You may build your own keys. There are plenty of examples for that in rules.d/, for example `60-evdev.rules` and `60-sensor.rules`

After some trials, it looks to work well. The customized key allows me to match from node (e.g. fw:1) as well from unit (e.g. fw1.0) and it's convenient, Later, I update the remote branch with revised commits. Thanks for your advice!